Make ExUnit server timeout configurable

JGulbronson opened the pull request:
Fixes https://github.com/elixir-lang/elixir/issues/5110

Qqwy commented on lib/ex_unit/lib/ex_unit/server.ex:
```diff
@@ -21,11 +20,13 @@ defmodule ExUnit.Server do
 
   def take_async_cases(count) do
-    GenServer.call(__MODULE__, {:take_async_cases, count}, @timeout)
+    timeout = Application.get_env(:ex_unit, :timeout)
```
What happens when this key has not been set in the application's environment? I think you might want the three-argument version of `Application.get_env` here.

josevalim commented:
Thank you @JGulbronson!

I would like to use another timeout then the one used for tests. Maybe we should call it `:case_timeout` which is the time taken to load a case. In order to do that, we would need to:

1. Document the new option
2. Add a default of `60_000` in lib/ex_unit/mix.exs
3. Change the code to use the new option

Could you go ahead with these changes please?

whatyouhide commented:
Would the name `:case_load_timeout` be ok as well? It seems a bit clearer than `:case_timeout` :)

josevalim commented:
@whatyouhide Yes, that's better!

josevalim commented:
@JGulbronson ping.

JGulbronson commented:
Sorry for the delay, was traveling this weekend. Seems like I misunderstood the original issue, but have a better understanding now. Should have a PR up by tomorrow night at the latest!

JGulbronson commented:
Okay, I believe this is more what you were looking for. I've introduced a new option, `:case_load_timeout`, which has a default timeout of 60 seconds, added it to the docs, and replaced the two @timeout calls with it.

lexmag commented on lib/ex_unit/test/ex_unit/runner_test.exs:
```diff
@@ -19,7 +19,7 @@ defmodule ExUnit.RunnerTest do
     refute config.trace
-    assert config.max_cases == :erlang.system_info(:schedulers_online) * 2
     assert config.timeout == 60_000
+    assert config.max_cases == :erlang.system_info(:schedulers_online) * 2
```
I think we don't need this change.

josevalim replied:
I think @JGulbronson wanted to test the case_load_timeout instead?

lexmag commented on lib/ex_unit/lib/ex_unit.ex:
```diff
@@ -180,6 +180,9 @@ defmodule ExUnit do
       `@tag capture_log: false`. Defaults to `false`.
 
+    * `:case_load_timeout` - the timeout to be used when loading a test case.
+      Defaults to 60_000ms
```
Missing period after `60_000ms`.

whatyouhide replied:
Also, maybe we can format this in a nicer way by using backticks:

```
Defaults to `60_000` milliseconds.
```

I think it reads a bit cleaner :)

JGulbronson commented:
Apologies for the delay, am in the middle of packing for going back to school. I've pushed a new commit, where tests pass locally. Gonna keep an eye on Travis, and make sure it goes green.

lexmag commented on lib/ex_unit/test/ex_unit/runner_test.exs:
```diff
@@ -19,7 +19,6 @@ defmodule ExUnit.RunnerTest do
     {_, config} = ExUnit.Runner.configure(opts)
 
     refute config.trace
-    assert config.max_cases == :erlang.system_info(:schedulers_online) * 2
```
I think we need this test back.

lexmag commented on lib/ex_unit/lib/ex_unit/server.ex:
```diff
@@ -21,11 +20,13 @@ defmodule ExUnit.Server do
 
   def take_async_cases(count) do
-    GenServer.call(__MODULE__, {:take_async_cases, count}, @timeout)
+    timeout = Application.get_env(:ex_unit, :case_load_timeout)
```
I would be better to use `fetch_env!` instead of `get_env` here and below.

josevalim commented:
Merged manually, thank you!

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

y86 commented:
Hi. I've had the same issue op had regarding this timeout which led me here. I tried looking for this config option on the current version of `lib/ex_unit/lib/ex_unit/server.ex`, but what I see there is the hard coded `@timeout`.

Has this config been merged or rolled back? Is there another way to accomplish the same thing on newer versions?

josevalim commented:
The timeout is set to infinity now, so you shouldn't be having issues.

y86 commented:
I am sorry about the confusion. I reviewed version 1.9 docs and found the correct option name.
Thanks!

davidlt commented:
Is there a way to set global timeout before I call "make test"? I am hitting "** (ExUnit.TimeoutError) test timed out after 60000ms." with 1.11.2.

```
  1) test in/2 too large list in guards (KernelTest)
     test/elixir/kernel_test.exs:429
     ** (ExUnit.TimeoutError) test timed out after 60000ms. You can change the timeout:

  2) test properly formats errors (Kernel.CLITest)
     test/elixir/kernel/cli_test.exs:46
     ** (ExUnit.TimeoutError) test timed out after 60000ms. You can change the timeout:

  3) test properly formats errors (Kernel.CLI.RPCTest)
     test/elixir/kernel/cli_test.exs:85
     ** (ExUnit.TimeoutError) test timed out after 60000ms. You can change the timeout:

  4) test encode then decode is identity (BaseTest)
     test/elixir/base_test.exs:783
     ** (ExUnit.TimeoutError) test timed out after 60000ms. You can change the timeout:

  5) test defguard(p) compilation fails on expressions not allowed in guards (Kernel.GuardTest)
     test/elixir/kernel/guard_test.exs:272
     ** (ExUnit.TimeoutError) test timed out after 60000ms. You can change the timeout:
```

josevalim commented:
Yes, if you look at the error message printed on each test it should tell you exactly how to configure it. :)

davidlt commented:
None of those options seems to apply directly to "make test". Looking at Makefile I don't see a way to pass global timeout value from environment or something. I am trying to avoid modifying the code directly to increase the timeouts.

josevalim commented:
Oh, those are all in Elixir's test suite itself. I see. Those tests should definitely not take more than 60 seconds, so I would say there is something wrong going on. But yeah, you would have to change the test_helper.exs by hand... we can provide an environemnt variable, but I am afraid it may not solve the issue.

davidlt commented:
It's QEMU riscv64 thus it's slow :)

josevalim commented:
I see, if you send a PR, we can read things such as ELIXIR_TEST_TIMEOUT and ELIXIR_TEST_MAX_CASES to reduce the concurency and increase the timeout. :)
